File: docs/test-cases/mass-assignment.java
// Test case: mass-assignment (API8:2023)
// Spring Boot example of mass assignment via BeanUtils
package com.example.demo.controller;

import org.springframework.beans.BeanUtils;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/users")
public class UserController {

    @PostMapping
    public User createUser(@RequestBody UserDto dto) {
        User entity = new User();
        // BUG: copies ALL matching fields including role, isAdmin
        BeanUtils.copyProperties(dto, entity);
        return userRepository.save(entity);
    }

    @PutMapping("/{id}")
    public User updateUser(@PathVariable Long id, @RequestBody UserDto dto) {
        User entity = userRepository.findById(id).orElseThrow();
        // BUG: no field filter -- attacker adds "role":"ADMIN" to JSON body
        BeanUtils.copyProperties(dto, entity);
        return userRepository.save(entity);
    }

    // UserDto accepts any field the client sends, including privileged ones
    public static class UserDto {
        private String username;
        private String email;
        private String role;      // BUG: should not be settable from input
        private Boolean isAdmin;  // BUG: should not be settable from input
    }
}

File: docs/test-cases/mass-assignment.go
// Test case: mass-assignment (API8:2023)
// GORM example of mass assignment from decoded JSON
package main

import (
	"encoding/json"
	"net/http"

	"gorm.io/gorm"
)

type User struct {
	gorm.Model
	Username string `json:"username"`
	Email    string `json:"email"`
	Role     string `json:"role"`      // privileged field
	IsAdmin  bool   `json:"is_admin"`  // privileged field
}

func CreateUser(db *gorm.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var user User
		// BUG: decoding full request body into the ORM model
		// attacker sends {"username":"joe","role":"admin","is_admin":true}
		json.NewDecoder(r.Body).Decode(&user)
		// BUG: all decoded fields including Role and IsAdmin are persisted
		db.Create(&user)
		w.WriteHeader(http.StatusCreated)
	}
}
